---
title: "BF528 Project 3 Report - ATAC Sequencing"
author: "Dylan Tamayo"
date: "2024-05-01"
output: html_document
---

```{r setup, message=FALSE, warning=FALSE, echo=FALSE}
library(tidyverse)
library(ggrepel)
library(patchwork)
library(grid)
library(png)
```

**Methods and Quality Control**

For analysis, default parameters were used unless stated otherwise.

GRCh38 was used as the reference genome, and the genome index was constructed using Bowtie2 v2.5.3 (Langmead & Salzberg, 2012) using default parameters.

Illumina adapters and low quality bases were then trimmed from sample reads using trimmomatic v0.39 (Bolger et al., 2014) in paired end mode. The Nextera Transposase Sequence adapters were removed using trimmomatic with the following setting: ILLUMINACLIP:NexteraPE-PE.fa:2:30:10. The following were also removed using trimmomatic: leading and trailing bases with a Phred score below 3; and reads with an average per-base quality below 15 in a 4-base wide sliding window.

Quality control of original and trimmed samples was conducted using FastQC v0.12.1-0 (https://www.bioinformatics.babraham.ac.uk/projects/fastqc/).

Paired sample reads were then aligned to GRCh38 using bowtie2 v2.5.3. Using samtools v1.19.2 (Danecek et al., 2021), alignments were converted to BAM files, sorted, and indexed. Alignments falling on the mitochondrial chromosome were also removed using samtools.

Quality control of read alignment was conducted using the samtools flagstat utility and multiqc v1.20 (Ewels et al., 2016). Reads were then shifted to account for the tagmentation process using the alignmentSieve utility and ATACshift option of deeptools v3.5.4 (Ram√≠rez et al., 2016). Fragment size quality control was conducted using ATACseqQC v1.24.0 in R v4.3.1 (Ou et al., 2018).

BED files of called peaks for each replicate were generated using MACS3 v3.0.1 (Zhang et al., 2008). A list of reproducible peaks was then generated by intersecting the two BED files using bedtools v2.31.1 (Quinlan et al., 2010); an intersection was defined as at least 50% overlap of peaks between the two replicates. Signal-artifact regions were then removed from the intersected peaks with bedtools, using a list of hg38 regions commonly enriched across ATAC-seq experiments.

Genomic features of the filtered list of peaks was annotated according to the GENCODE GRCh38 primary assembly annotation v45 using HOMER v4.11 (Heinz et al., 2010). Enriched motifs at peak locations were also identified with HOMER, using the GENCODE GRCh38 human primary assembly genome.

To generate signal coverage plots for nucleosome-free regions (NFR) and nucleosome-bound regions (NBR), bigWig files were generated by first separating 0-100 bp (NFR) and 180-247 bp (NBR) fragments from the ATAC-shifted BAM files using the deeptools alignmentSieve utility, and then converted using the deeptools bamCoverage utility.

Transcription start and termination sites for genes for the GRCh38 genome were retrieved from the UCSC Table Browser; combining the TSS and TTS information with the NFR and NBR bigWig files for each replicate, a matrix of signal coverage was scored using the DeepTools computeMatrix utility in reference-point mode. Counts were calculated within a 2kb window up- and downstream of the TSS and TTS, respectively. Signal coverage was visualized for NFRs and NBRs for each replicate using the deeptools plotProfile utility.
<br>
<br>

**Quality Control Analysis**

Quality of reads was high overall prior to trimming, with almost all sequences having a Phred score greater than 30. GC content was also distributed normally. Nextera transposase adapters were present in the initial samples, and were trimmed out in the subsequent pre-processing step.

A small number of unpaired reads from each replicate, largely representing adapter sequences, were separated in the trimmomatic step and were excluded from further analysis.

According to the samtools flagstat utility, both replicates had high-quality alignment and mapping results, with a high percentage of mapped reads (~99.2%) and properly paired reads (~92%), as well as a low percentage of singletons (~0.25%) and reads with mates mapped to different chromosomes.

Around 25% of alignments remained in both replicates after removing alignments falling on the mitochondrial chromosome.

```{r, echo=FALSE}
# Define your data
step <- c("Before mitonchodria removal", "After mitonchodria removal")
ATACrep3 <- c(144081590, 36224722)
ATACrep4 <- c(106544916, 27551934)
data <- data.frame(step, ATACrep3, ATACrep4)

# Print the table using kable
knitr::kable(
  data,
  format = "html",
  table.attr = 'style="width: 50%; border: 1px solid black; padding: 20px;"',
  format.args = list(big.mark = ','),
  col.names = c("", "ATACrep3", "ATACrep4"),
  row.names = FALSE,
  caption = 'Table 1. Alignment quantities before and after removing mitochondrial chromosome alignments.',
)
```
<br>

Fragment sizes of aligned sequences were of high quality, with a tall peak below 100 bp, representing NFRs, and a periodicity in the curve representing integer multiples of NBRs.

```{r, echo=FALSE, message=FALSE}
# Define image paths
image_paths <- c(
  "results/ATACseqQC/ATACrep3_lib_complexity.png",
  "results/ATACseqQC/ATACrep4_lib_complexity.png",
  "results/ATACseqQC/ATACrep3_frag_size.png",
  "results/ATACseqQC/ATACrep4_frag_size.png"
)

# Load images and arrange in a grid
grid.newpage()
pushViewport(viewport(layout = grid.layout(2, 2)))
for (i in seq_along(image_paths)) {
  img <- readPNG(image_paths[i])
  grid.raster(img, vp = viewport(layout.pos.row = (i - 1) %/% 2 + 1, layout.pos.col = (i - 1) %% 2 + 1))
}
```
<br>
<strong>
Top: Library size complexity for both replicates.

Bottom: Fragment size distribution for both replicates.
</strong>
<br>
<br>

**Signal Coverage**

Signal coverage was similar between replicates for both the nucleosome-free and nucleosome-bound regions. 

```{r, echo=FALSE, message=FALSE}
# Define image paths
image_paths <- c(
  "results/signal_coverage_ATACrep3_0_100.png",
  "results/signal_coverage_ATACrep4_0_100.png",
  "results/signal_coverage_ATACrep3_180_247.png",
  "results/signal_coverage_ATACrep4_180_247.png"
)

# # Load images into a grid
# image_grid <- lapply(image_paths, function(path) {
#   knitr::include_graphics(path)
# })

# Load images and arrange in a grid
grid.newpage()
pushViewport(viewport(layout = grid.layout(2, 2)))
for (i in seq_along(image_paths)) {
  img <- readPNG(image_paths[i])
  grid.raster(img, vp = viewport(layout.pos.row = (i - 1) %/% 2 + 1, layout.pos.col = (i - 1) %% 2 + 1))
}
```
<br>
<strong>
Top: Signal coverage for nucleosome-free regions for both replicates.

Bottom: Signal coverage for nucleosome-bound regions for both replicates.
</strong>
<br>
<br>

**Peak Calling, Motifs and Annotation**

To produce a list of reproducible peaks, the two BED files were intersected, keeping peaks that overlapped by at least 50%. Peaks that are present in both replicates are more likely to be reproduced in other samples over one-off peaks present in only one sample. Requiring an overlap of 50% provided flexibility for peaks to be slightly off between replicates.

While both replicates had over 51,000 peaks called by MACS3, only 502 peaks were present in both samples, with 481 peaks occurring in non-blacklisted regions. One peak had no matching annotations through HOMER, leaving 480 annotated peaks remaining for further analysis.

```{r, echo=FALSE}
# Define your data
bam <- c('ATACrep3', 'ATACrep4', 'Intersect', 'Remove blacklist')
peaks <- c(58893, 51137, 502, 481)
table2 <- data.frame(bam, peaks)

# Print the table using kable
knitr::kable(
  table2,
  format = "html",
  table.attr = 'style="width: 50%; border: 1px solid black; padding: 20px;"',
  format.args = list(big.mark = ','),
  col.names = c('BED File', 'Peaks'),
  row.names = FALSE,
  caption = 'Table 2. Quantities of peaks called.',
)
```
<br>
<br>
The majority of peaks annotated by HOMER were found near introns, promoters, or intergenic regions, with the remaining peaks located near exons or transcription termination sites. 

```{r, echo=FALSE}
pattern <- '^(.*?)(?=\\s*\\([^()]*\\))|(\\w+)$'

annot <- read_tsv('results/ATAC_annotated.bed',
                  show_col_types = FALSE) %>%
  # replace_na(list(Annotation = 'NA')) %>
  drop_na(Annotation) %>% 
  mutate(clean_annot = str_extract(Annotation, pattern),
         clean_annot = case_when(clean_annot == 'NA' ~ 'N/A',
                                 clean_annot == 'Intergenic' ~ 'intergenic',
                                 .default = clean_annot))
annot2 <- annot %>% group_by(clean_annot) %>% count()

annot3 <- annot2 %>% mutate(perc = n/sum(annot2$n)*100,
         percf = round(perc, 1),
         perb = str_c(percf, '%'))

bar_chart <- ggplot(annot3, aes(x = clean_annot, y = n, fill = clean_annot)) +
  geom_col() +
  labs(x = 'Peak Annotation', y = 'Number of Peaks') +
  theme_bw() +
  geom_text(data = annot2,
            aes(label = n, y = n),
            position = position_fill(),
            vjust = -0.3) +
  scale_x_discrete(guide = guide_axis(angle = 45)) +
  theme(legend.position="none")
  
pie_chart <- ggplot(annot3, aes(x = '', y = perc, fill = clean_annot)) +
  geom_col() +
  coord_polar(theta = "y") +
  labs(fill = 'Peak Annotation') +
  geom_text_repel(aes(label = perb),
            position = position_stack(vjust = 0.5)) +
  theme_void()

combined_chart <- bar_chart + pie_chart +
  plot_annotation(title = 'Proportion of Accessible Chromatin Regions',
                  theme = theme(plot.title = element_text(hjust = 0.5)))
plot(combined_chart)
```
<br>
<strong>
Bar and pie chart showing proportions of annotated regions associated with peaks.
</strong>
<br>

Analyzing the known motifs found through HOMER, the top ranking motif was for CCCTC-binding factor (CTCF), which is a protein that regulates chromatin architecture (Ong et al., 2014). The second-rank motif was for a CTCF paralog and oncogene, BORIS, which is highly expressed in ovarian cancer.

Other top motifs included a motif for Jurkat cells, an immortalized line of human T lymphocyte cells isolated from an individual with leukemia (Gioia et al., 2018), as well as a factor involved in the development of dendritic cells (Grajales-Reyes et al., 2015).

```{r, echo=FALSE}
knitr::include_graphics("results/ATAC_motif_results.png")
```
<br>
<strong>
Top five results for motifs found using HOMER.
</strong>
<br>
<br>
To analyze gene enrichment, the 158 gene symbol annotations of promoter-associated peaks were analyzed on the Enrichr website (https://maayanlab.cloud/Enrichr/). Using the Azimuth Cell Types 2021 database, which provides cell type identification for ATAC-seq experiments (https://azimuth.hubmapconsortium.org/), the top results for with adjusted p-value < 0.05 related to B cells. Connected genes such as BANK1 and MEF2C are involved in B cell receptor signaling pathways (Corinaldesi et al., 2022).

```{r, echo=FALSE}
knitr::include_graphics("results/Azimuth_Cell_Types_2021_bar_graph.png")
knitr::include_graphics("results/clustergrammer.png")

enrichr_res <- read_tsv('results/Azimuth_Cell_Types_2021_table.tsv',
                        show_col_types = FALSE) %>% 
  select(c(Term, Overlap, `Adjusted P-value`, `Odds Ratio`, Genes))
  
print(head(enrichr_res, n = 10))
```
<strong>
Enrichr results for annotated genes.
</strong>
<br>
<br>

**Discussion**

Based on the gene annotations of the peaks as well as the peak motifs, it is likely that samples originate from some immune cell line, given the presence of B cell-related genes from the peak annotations, as well as dendritic cells in the motifs. As the motifs also contain the BORIS oncogene and T cells related to leukemia, the cell line may be cancerous as well.

The presence of CTCF as the top motif provides evidence that the ATACseq procedure isolated accessible chromatin, given the role CTCF plays in facilitating interactions with transcription factors and chromosomal organization. 

Further directions for analysis include analyzing motifs and enriched genes for NFRs and NBRs separately, to assess differences in expression based on accessible chromatin. Other directions include integration of ATACseq data with analyses such as single-cell RNA sequencing and ChIP-seq, which can provide additional information on the genetic expression and identity of the original cell line.
<br>
<br>

**References**

Bolger, Anthony M., Marc Lohse, and Bjoern Usadel. "Trimmomatic: a flexible trimmer for Illumina sequence data." <i>Bioinformatics</i> 30.15 (2014): 2114-2120.

Corinaldesi, Clarissa, et al. "Tracking immunoglobulin repertoire and transcriptomic changes in germinal center B cells by single-cell analysis." <i>Frontiers in Immunology</i> 12 (2022): 818758.

Danecek, Petr, et al. "Twelve years of SAMtools and BCFtools." <i>Gigascience</i> 10.2 (2021): giab008.

Ewels, Philip, et al. "MultiQC: summarize analysis results for multiple tools and samples in a single report." <i>Bioinformatics</i> 32.19 (2016): 3047-3048.

Gioia, Louis, et al. "A genome-wide survey of mutations in the Jurkat cell line." <i>BMC genomics</i> 19 (2018): 1-13.

Grajales-Reyes, Gary E., et al. "Batf3 maintains autoactivation of Irf8 for commitment of a CD8Œ±+ conventional DC clonogenic progenitor." <i>Nature immunology</i> 16.7 (2015): 708-717.

Heinz, Sven, et al. "Simple combinations of lineage-determining transcription factors prime cis-regulatory elements required for macrophage and B cell identities." <i>Molecular cell</i> 38.4 (2010): 576-589.

Ong, Chin-Tong, and Victor G. Corces. "CTCF: an architectural protein bridging genome topology and function." <i>Nature Reviews Genetics</i> 15.4 (2014): 234-246.

Ou, Jianhong, et al. "ATACseqQC: a Bioconductor package for post-alignment quality assessment of ATAC-seq data." <i>BMC genomics</i> 19 (2018): 1-13.

Quinlan, Aaron R., and Ira M. Hall. "BEDTools: a flexible suite of utilities for comparing genomic features." <i>Bioinformatics</i> 26.6 (2010): 841-842.

Ram√≠rez, Fidel, et al. "deepTools2: a next generation web server for deep-sequencing data analysis." <i>Nucleic acids research</i> 44.Web Server issue (2016): W160.

Xie, Zhuorui, et al. "Gene set knowledge discovery with Enrichr." <i>Current protocols</i> 1.3 (2021): e90.

Zhang, Yong, et al. "Model-based analysis of ChIP-Seq (MACS)." <i>Genome biology</i> 9 (2008): 1-9.
<br>
